crates/sui-core/src/authority/authority_store.rs

```diff
@@ -897,18 +882,20 @@ impl AuthorityStore {
         owned_input_objects: &[ObjectRef],
         tx_digest: TransactionDigest,
     ) -> SuiResult {
-        // Other writers may be attempting to acquire locks on the same objects, so a mutex is
+        // Other writers may be attempting to acquire locks on the same objects, so a transaction is
         // required.
-        // TODO: replace with optimistic transactions (i.e. set lock to tx if none)
-        let _mutexes = self.acquire_locks(owned_input_objects).await;
+        let transaction = self
+            .perpetual_tables
+            .owned_object_transaction_locks
+            .transaction()?;
 
         debug!(?tx_digest, ?owned_input_objects, "acquire_locks");
         let mut locks_to_write = Vec::new();
 
-        let locks = self
-            .perpetual_tables
-            .owned_object_transaction_locks
-            .multi_get(owned_input_objects)?;
+        let locks = transaction.multi_get(
```

Comment:
I believe without a `multi_get_for_update` this code is incorrect and race-prone. See https://github.com/MystenLabs/sui/pull/7044/files#diff-8d247172a1f7c5ee07cbbf685d8388b0b9d97eb81009ba4d1c006c41aa8d5f86R666 from my pending PR for a test case that shows how this can fail.

I think this strengthens the case for making `transaction_with_snapshot()` the default option, as it is correct by default.

Reply:
Also I think any transaction needs to be wrapped in a retry loop. (also see my PR for a retry macro). Otherwise we will expose the transaction errors to the client.

Reply:
yes, that's why I've changed it to draft for now. Will either convert to multi_get_for_update or transaction with snapshot

Reply:
sorry, I did see it was a draft, but since you assigned me already I figured i'd review it 😄